module/zfs/zfs_vfsops.c

```diff
@@ -1769,6 +1769,12 @@ zfs_remount(struct super_block *sb, int *flags, zfs_mnt_t *zm)
 	if (error)
 		return (error);
 
+	if (dmu_objset_is_snapshot(zfsvfs->z_os) && !vfsp->vfs_readonly) {
+		ASSERT(vfsp->vfs_do_readonly);
+		*flags |= MS_RDONLY;
+		return (EROFS);
+	}
```

Comment:
The readonly flags is always exists in flags as MS_RDONLY. We cannot rely on the option string for common mount flags.
Also, we should do the same thing for reaonly import.